### PR TITLE
[feat]: [CDS-51183]: add fetch organizations and workspaces endpoints

### PR DESCRIPTION
**Pull request by tmacari-harness**
## Describe your changes
[feat]: [CDS-51183]: add fetch organizations and workspaces endpoints
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
    - CodeFormat: `trigger codeformat`
    - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

[CDS-51183]: https://harness.atlassian.net/browse/CDS-51183?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/43129)

### CODE REVIEW

**Comment by tmacari-harness** [comment]
trigger messagecheck

**Comment by tmacari-harness** [comment]
trigger ft1

File: 932-connector-task/src/main/java/io/harness/connector/task/terraformcloud/TerraformCloudValidationHandler.java [hunk 1]
```diff
@@ -37,10 +49,28 @@ public ConnectorValidationResult validate(
...
-    return ConnectorValidationResult.builder()
-        .status(ConnectivityStatus.SUCCESS)
-        .testedAt(System.currentTimeMillis())
-        .build();
+    TerraformCloudApiTokenCredentials terraformCloudApiTokenCredentials =
+        (TerraformCloudApiTokenCredentials) terraformCloudConfig.getTerraformCloudCredentials();
+    ConnectorValidationResultBuilder result = ConnectorValidationResult.builder();
+    try {
+      TerraformCloudResponse<List<OrganizationData>> organizationsResponse = terraformCloudClient.listOrganizations(
+          terraformCloudApiTokenCredentials.getUrl(), terraformCloudApiTokenCredentials.getToken(), 1);
```
Comment (vlica-harness) [inline 932-connector-task/src/main/java/io/harness/connector/task/terraformcloud/TerraformCloudValidationHandler.java:57]: should we always have page 1, in the last parameter, or this means number of pages, if number of pages would be better to rename in api, and update to be a constant?

File: 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  private final TerraformCloudResourceService terraformCloudResourceService;
+  private final OrgAndProjectValidationHelper orgAndProjectValidationHelper;
+
+  @GET
+  @Path("organizations")
+  @ApiOperation(value = "Gets terraform cloud organizations", nickname = "getTerraformCloudOrganizations")
+  public ResponseDTO<OrganizationsDTO> getOrganizations(
+      @QueryParam("connectorRef") String terraformCloudConnectorIdentifier,
+      @NotNull @QueryParam(NGCommonEntityConstants.ACCOUNT_KEY) String accountId,
+      @QueryParam(NGCommonEntityConstants.ORG_KEY) String orgIdentifier,
```
Comment (vlica-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java:61]: I think orgIdentifier and projectIdentifier should have also @notNull

File: 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @GET
+  @Path("workspaces")
+  @ApiOperation(value = "Gets terraform cloud workspaces", nickname = "getTerraformCloudWorkspaces")
+  public ResponseDTO<WorkspacesDTO> getWorkspaces(@QueryParam("connectorRef") String terraformCloudConnectorIdentifier,
+      @NotNull @QueryParam(NGCommonEntityConstants.ACCOUNT_KEY) String accountId,
+      @QueryParam(NGCommonEntityConstants.ORG_KEY) String orgIdentifier,
+      @QueryParam(NGCommonEntityConstants.PROJECT_KEY) String projectIdentifier,
+      @QueryParam("organization") String organization) {
```
Comment (vlica-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java:77]: if organization is mandatory, can we have @notnull and @notempty

File: 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+import javax.ws.rs.Produces;
+import javax.ws.rs.QueryParam;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+@OwnedBy(CDP)
+@Api("terraform-cloud")
+@Path("/terraform-cloud")
+@Produces({"application/json"})
```
Comment (vlica-harness) [inline 120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java:42]: minor: I think we can add as well application/yaml"

**Review by tmacari-harness — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129)* status: 36 file(s) to review. 5. unresolved discussions. 0 of 2 approvals obtained. 

___
*[`120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java` line 42 at r1](https://reviewable.io/reviews/harness/harness-core/43129#-NMmXSyAqW7MLj6fj07O-r1-42:-NN12x_c96Qcj5vkxidL:b-xjvl9q) ([raw file](https://github.com/harness/harness-core/blob/09b3dad57844a6a8a74205b8ee03c3fd8ab86254/120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java#L42)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

minor: I think we can add as well application/yaml"
</blockquote></details>

ti would never return YAML



_Code quote:_
```Java
 public ResponseDTO<OrganizationsDTO> 
```
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java` line 61 at r1](https://reviewable.io/reviews/harness/harness-core/43129#-NMmXSyAqW7MLj6fj07O-r1-61:-NN131NG5MteT4f6Sn5V:b-8o5hdk) ([raw file](https://github.com/harness/harness-core/blob/09b3dad57844a6a8a74205b8ee03c3fd8ab86254/120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java#L61)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

I think orgIdentifier and projectIdentifier should have also @notNull
</blockquote></details>

Good point, done
___
*[`120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java` line 77 at r1](https://reviewable.io/reviews/harness/harness-core/43129#-NMmXSyAqW7MLj6fj07O-r1-77:-NN13UN88t0Kr4dgaQ3u:bb6un1j) ([raw file](https://github.com/harness/harness-core/blob/09b3dad57844a6a8a74205b8ee03c3fd8ab86254/120-ng-manager/src/main/java/io/harness/ng/core/terraformcloud/resources/TerraformCloudResource.java#L77)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

if organization is mandatory, can we have @notnull and @notempty
</blockquote></details>

Done
___
*[`932-connector-task/src/main/java/io/harness/connector/task/terraformcloud/TerraformCloudValidationHandler.java` line 57 at r1](https://reviewable.io/reviews/harness/harness-core/43129#-NMmXSyAqW7MLj6fj07e-r1-57:-NN13xyR1muNiPTeLb8z:b-b8lhx7) ([raw file](https://github.com/harness/harness-core/blob/09b3dad57844a6a8a74205b8ee03c3fd8ab86254/932-connector-task/src/main/java/io/harness/connector/task/terraformcloud/TerraformCloudValidationHandler.java#L57)):*
<details><summary><i>Previously, vlica-harness (Vasile Lica) wrote…</i></summary><blockquote>

should we always have page 1, in the last parameter, or this means number of pages, if number of pages would be better to rename in api, and update to be a constant?  
</blockquote></details>

It's page number, we need 1 only here, I don't think we have to make it a constant.

**Review by vlica-harness — APPROVED** [review]
> Reviewed 35 of 36 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129)* status: 36 file(s) to review. 5. unresolved discussions. 1 of 2 approvals obtained.

**Review by alebuh — APPROVED** [review]
> Reviewed 35 of 36 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129)* status: 5. unresolved discussions.

**Review by anilchowdhury-harness — APPROVED** [review]
:lgtm:

> Reviewed 35 of 36 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129#-:-NN1XdT8BhVZnljD1tlz:bnfp4nl)* status: 5. unresolved discussions.

**Review by vlica-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129)* status: :shipit: complete! ✓

**Review by mdmilic — APPROVED** [review]
Approve delegate owned files

> Reviewed 12 of 36 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/43129#-:-NN1gbgi7skx9Hl-lTGK:b-nvcshz)* status: :shipit: complete! ✓

**Pull request merged**